# 4.7.5 Naive Bayes

Next, we fit a naive Bayes model to the `Smarket` data. Naive Bayes is implemented in `R` using the `naiveBayes()` function, which is part of the `e1071` library. The syntax is identical to that of `lda()` and `qda()`. By default, this implementation of the naive Bayes classifier models each quantitative feature using a Gaussian distribution. However, a kernel density method can also be used to estimate the distributions.

In [1]:
library(ISLR2)
library(e1071)
attach(Smarket)
train <- (Year < 2005)
Smarket.2005 <- Smarket[!train,]
Direction.2005 <- Direction[!train]

In [2]:
nb.fit <- naiveBayes(Direction ~ Lag1 + Lag2, data = Smarket, subset = train)
nb.fit


Naive Bayes Classifier for Discrete Predictors

Call:
naiveBayes.default(x = X, y = Y, laplace = laplace)

A-priori probabilities:
Y
    Down       Up 
0.491984 0.508016 

Conditional probabilities:
      Lag1
Y             [,1]     [,2]
  Down  0.04279022 1.227446
  Up   -0.03954635 1.231668

      Lag2
Y             [,1]     [,2]
  Down  0.03389409 1.239191
  Up   -0.03132544 1.220765


The output contains the estimated mean and standard deviation for each variable in each class. For example, the mean for `Lag1` is $0.0428$ for `Direction=Down`, and the standard deviation is $1.23$. We can easily verify this:

In [3]:
mean(Lag1[train][Direction[train] == "Down"])

[1] 0.04279022

In [4]:
sd(Lag1[train][Direction[train] == "Down"])

[1] 1.227446

The `predict()` function is straightforward.

In [5]:
nb.class <- predict(nb.fit, Smarket.2005)
table(nb.class, Direction.2005)

        Direction.2005
nb.class Down  Up
    Down   28  20
    Up     83 121

In [6]:
mean(nb.class == Direction.2005)

[1] 0.5912698

Naive Bayes performs very well on this data, which accurate predictions over $59%$ of the time. This is slightly worse than QDA, but much better than LDA.  

The `predict()` function can also generate estimates of the probability that each observation belongs to a particular class.

In [7]:
nb.preds <- predict(nb.fit, Smarket.2005, type = "raw")
nb.preds[1:5,]

Down      Up       
[1,] 0.4873164 0.5126836
[2,] 0.4762492 0.5237508
[3,] 0.4653377 0.5346623
[4,] 0.4748652 0.5251348
[5,] 0.4901890 0.5098110